Lib/http/client.py

```diff
@@ -941,23 +941,26 @@ def _tunnel(self):
         del headers
 
         response = self.response_class(self.sock, method=self._method)
-        (version, code, message) = response._read_status()
+        try:
+            (version, code, message) = response._read_status()
 
-        if code != http.HTTPStatus.OK:
-            self.close()
-            raise OSError(f"Tunnel connection failed: {code} {message.strip()}")
-        while True:
-            line = response.fp.readline(_MAXLINE + 1)
-            if len(line) > _MAXLINE:
-                raise LineTooLong("header line")
-            if not line:
-                # for sites which EOF without sending a trailer
-                break
-            if line in (b'\r\n', b'\n', b''):
-                break
+            if code != http.HTTPStatus.OK:
+                self.close()
+                raise OSError(f"Tunnel connection failed: {code} {message.strip()}")
+            while True:
+                line = response.fp.readline(_MAXLINE + 1)
+                if len(line) > _MAXLINE:
+                    raise LineTooLong("header line")
+                if not line:
+                    # for sites which EOF without sending a trailer
+                    break
+                if line in (b'\r\n', b'\n', b''):
+                    break
 
-            if self.debuglevel > 0:
-                print('header:', line.decode())
+                if self.debuglevel > 0:
+                    print('header:', line.decode())
+        finally:
+            response.close()
```

Comment:
This behavior should be similar to `getreponse` above. I assume getresponse should handle the close() and I am uncertain why urllib3 ran into this here.

Reply:
getresponse keeps a reference to `self.__response` so the most recent response can be closed when the connection is closed

Reply:
getresponse doesn't handle the close, it's intended to be called by consumer of HTTPResponse

Reply:
> I am uncertain why urllib3 ran into this here.

This is a socket leak and I'm adding the infrastructure to urllib3 to detect socket leaks and it found this one

Usually you can rely on `response.__del__` calling `response.close` but that's prevented when `_tunnel` raises an `OSError` or `TimeoutError`